### PR TITLE
Twri

### LINKED ISSUE #36: Born modeling J*dm on twri branch
Hi,

I am now at the latest version of twri and the born modeling seems not working. My code is here [https://github.gatech.edu/zyin62/multiples/blob/master/MFE.jl](url). I would really appreciate your help. Thanks!

Francis

### PR DESCRIPTION
**Pull request by mloubout**
This PR is (once again) quite massive but ships quite a few nice features (

# PhysicalParameter

Thre is a new structure type callled `PhysicalParameters`, that inherit from Julia's `AbstractVector` to represent physical domain arrays. These presents to Julia as vectors `(nx*ny*nz)` for Linear Algebra purpose but keep the n-dimensional array internally with its size, grid spacing, and origin. 

Each field is accessible via standard `p.o` (origin), `p.d` (spacing), `p.n` (nx, ny, nz) and `p.data` (nx x ny in 2D). All norms/addition/... are overloaded and rules are the same as Julia standard array.

```julia
p = PhysicalParameter(randn(10, 10), n, d, o) # n is optional here, but mandatory if the input is a vector

p .+ 1 # adds 1 to everypoint
p + 1 # This will error same ways as standard vector, need broadcasting

2 * p # multiply by 2
....
```

Additionally, these support (beta) multiplication by standard matrices and JOLI operators.

```
A = randn(length(p), length(p))
A*p
```

Added all the test with it. Documentation for it has been added as weel in `data_structures.md` :

**Notes**

Norms are currently still addimentional. The main reason is that adding space dimension to the data side is non=trivial and will need its own PR/update/.. To summarize the issue in two lines, consider the standard Jacobian dottest `< J dm, d> = <dm, J' d>`. It is trivial to add the spacing `h_x*h_y` to the RHS of the equation, however, for the LHS, this `h_x * h_y` would need to be added while none of the two terms (J dm, d) depend on the discretization explicitly.

**Notes2**

`PhysicalParameter` can be plotted directly in 2D via `imshow(p)`, for other sizes and slices, need to plot `p.data`

# Model update

The Model structure has been simplified (Model_TTI is gone) and now only contains `n, d, o` and a field called `params`that is a dictionary of `:symbol, PhysicalParameter` such as `:m` or `:epsilon` and corresponding array. Previous shortcuts such as `model.m` still work properly.

# TWRI

A `twri_objectiv(model, q, d, y; options=Options, optionswri=TWRIOpions)` is added and implements the math from the paper (pt arxiv ref once submitted).

TWRIOpions contains different compute options:
`grad_corr=false` alpha correction from Eq 21 (in combination with below)
`comp_alpha=false` computes optimal alpha Eq 19 
`weight_fun=nothing` whether to use focusing, input as tuple ("name", value) and supports (undocumentde for now) "srcfocus" and "depth" only.
`eps=0` noise level
`params=:m` Which params to compute gradient w.r.t, can be `:m, :y, :all, nothing`

Tested in `test_gradient_wri.jl` that does the gradient test:
- w.r.t m for fixed y with `comp_alpha` and `grad_corr` disabled
- w.r.t m for  `y= alpha * r` and corresponding optimal alpha and gradient correction (see paper)
- w.r.t y for fixed `m` with `comp_alpha` and `grad_corr` disabled

# Misc

Fixes #36 with 26275d4

### CODE REVIEW

File: docs/data_structures.md [hunk 1]
```diff
@@ -1,5 +1,42 @@
 # Data structures
 
+## Physical Parameter
+
+Data structure for physicall parameter array in JUDI. A `PhysicalParameter` inherits from julia `AbstractVector`
```
Comment (alisiahkoohi) [inline docs/data_structures.md:5]: small typo: physicall -> physical

File: examples/twri/data/gen_data_bg_acou.jl [hunk 2]
```diff
@@ -0,0 +1,57 @@
+################################################################################
+#
+# Generate synthetic data for the BG Compass model
+#
+# Author: Mathias Loubout and Gabbrio Rizzuti
+# Date: September 2020
+################################################################################
+### Module loading
+
+using JUDI, JLD2, JUDI.TimeModeling
```
Comment (alisiahkoohi) [inline examples/twri/data/gen_data_bg_acou.jl:10]: Just a general remark—here the user has to install JLD2 herself? I understand that we do no want to add unnecessary packages to the dependencies but on the other hand here the user will encounter an error most of the time.

Comment (mloubout) [reply examples/twri/data/gen_data_bg_acou.jl:10]: Well  two things
- (some) people are already complaining that there is too many dependencies. 
- Adding JLD2  to the dependencies would not fi the issue since JLD2  would be available to JUDI (`julia --project`) but not to the user.

But should add remarks or something to tell what people need for the examples.

File: examples/twri/scripts/BGCompass_inversion_tti.jl [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    return m
+end
+mmin = 1f0/4.5f0^2
+mmax = 1f0/1.4f0^2
+ProjBound(x) = proj_bounds(x, mmin,  mmax)
+
+for var=vvar
+    for smooth_thomsen=[true, false]
+        ### Background model
+        idx_w = 17
```
Comment (alisiahkoohi) [inline examples/twri/scripts/BGCompass_inversion_tti.jl:32]: can some of these variables made global? `idx_w`, `vare`, `fevals`, and `batchsize`.

File: examples/twri/scripts/BGCompass_inversion_tti.jl [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    m[m[:] .< mmin] .= mmin
+    m[m[:] .> mmax] .= mmax
+    return m
+end
+mmin = 1f0/4.5f0^2
+mmax = 1f0/1.4f0^2
+ProjBound(x) = proj_bounds(x, mmin,  mmax)
+
+for var=vvar
+    for smooth_thomsen=[true, false]
```
Comment (alisiahkoohi) [inline examples/twri/scripts/BGCompass_inversion_tti.jl:30]: Seems like `smooth_thomsen=[true, false]` can be replaced by `anis = ["st", "tt"]`

File: examples/twri/scripts/GaussLens_inversion.jl [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+#
+# Run inversion for the Gaussian lens model
+#
+################################################################################
+
+R = Float32
+
+### Module loading
+using LinearAlgebra, JLD2, PyPlot, Dates, Distributed
+using Optim, LineSearches
```
Comment (alisiahkoohi) [inline examples/twri/scripts/GaussLens_inversion.jl:11]: Are `Optim` and `LineSearches` in the dependencies or described in `twri/examples/README.md`?

File: examples/twri/scripts/GaussLens_inversion.jl [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+using JUDI, JUDI.TimeModeling, JUDI.SLIM_optim
+
+### Load synthetic data
+base_path = dirname(pathof(JUDI))*"/../examples/twri/"
+@load string(base_path*"data/GaussLens_data_acou.jld") model_true dat fsrc
```
Comment (alisiahkoohi) [inline examples/twri/scripts/GaussLens_inversion.jl:16]: A small note somewhere is needed that describes in which order the scripts should be run. Perhaps somewhere in `examples/twri`

File: src/TimeModeling/judiVector.jl [hunk 7]
```diff
@@ -288,37 +288,8 @@ adjoint(a::judiVector{vDT}) where vDT =
 
 ##########################################################
 
-# Overload base function for SegyIO objects
-
+# Overload needed base function for SegyIO objects
 vec(x::SegyIO.SeisCon) = vec(x[1].data)
-norm(x::SegyIO.IBMFloat32; kwargs...) = norm(convert(Float32,x); kwargs...)
```
Comment (alisiahkoohi) [inline src/TimeModeling/judiVector.jl:294]: Just curious, why aren't these needed anymore?

Comment (mloubout) [reply src/TimeModeling/judiVector.jl:294]: They were never needed, not used anywhere. And if it where these belongs in SegyIO not JUDI

File: examples/twri/scripts/GaussLens_inversion.jl (hunk 6 above)
Comment (mloubout) [reply examples/twri/scripts/GaussLens_inversion.jl:16]: Good point, replaced by a `try/catch` to generate data if doesn't exist.

File: examples/twri/scripts/GaussLens_inversion.jl (hunk 5 above)
Comment (mloubout) [reply examples/twri/scripts/GaussLens_inversion.jl:11]: Added to the README in `examples/README.md`

File: examples/twri/scripts/BGCompass_inversion_tti.jl (hunk 4 above)
Comment (mloubout) [reply examples/twri/scripts/BGCompass_inversion_tti.jl:30]: good point, done

File: examples/twri/scripts/BGCompass_inversion_tti.jl (hunk 3 above)
Comment (mloubout) [reply examples/twri/scripts/BGCompass_inversion_tti.jl:32]: yes, done

File: docs/data_structures.md (hunk 1 above)
Comment (mloubout) [reply docs/data_structures.md:5]: fixed

File: test/test_basics.jl [hunk 8]
```diff
@@ -36,16 +35,25 @@ function test_padding(ndim)
...
+    @test get_dt(model) == calculate_dt(model)
+    @test model[:m] == model.m
+    mloc = model.m
+    model.m .*= 0f0
+    @test norm(model.m) == 0
+    model.m = mloc
+    @test model.m == mloc
+
     srcGeometry = example_src_geometry()
     recGeometry = example_rec_geometry(cut=true)
```
Comment (alisiahkoohi) [inline test/test_basics.jl:58]: I know this line has not changed in this pull request but there are issues in this line when running the tests—i.e., `GROUP="All" julia --project test/runtests.jl`. Apparently `example_rec_geometry` is defined twice in [here](https://github.com/slimgroup/JUDI.jl/blob/twri/test/test_composite.jl#L5) and [here](https://github.com/slimgroup/JUDI.jl/blob/twri/test/utils.jl#L125), which to my understanding is causing the following error in testing:

```
Test basic utilities: Error During Test at /home/ali/.julia/dev/JUDI/test/test_basics.jl:114
  Got exception outside of a @test
  MethodError: no method matching example_rec_geometry(; cut=true)
  Closest candidates are:
    example_rec_geometry(; nsrc, nrec) at /home/ali/.julia/dev/JUDI/test/test_composite.jl:6 got unsupported keyword argument "cut"
  Stacktrace:
   [1] kwerr(::NamedTuple{(:cut,),Tuple{Bool}}, ::Function) at ./error.jl:157
   [2] test_limit_m(::Int64, ::Bool) at /home/ali/.julia/dev/JUDI/test/test_basics.jl:58
   [3] macro expansion at /home/ali/.julia/dev/JUDI/test/test_basics.jl:125 [inlined]
   [4] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1113 [inlined]
   [5] top-level scope at /home/ali/.julia/dev/JUDI/test/test_basics.jl:115
   [6] include(::String) at ./client.jl:439
   [7] macro expansion at ./util.jl:175 [inlined]
   [8] top-level scope at /home/ali/.julia/dev/JUDI/test/runtests.jl:52
   [9] include(::Module, ::String) at ./Base.jl:377
   [10] exec_options(::Base.JLOptions) at ./client.jl:288
   [11] _start() at ./client.jl:484
```

File: test/runtests.jl [hunk 9]
```diff
@@ -22,35 +23,43 @@ base = ["test_abstract_vectors.jl",
...
 devito = ["test_linearity.jl",
+          "test_adjoint.jl",
+          "test_all_options.jl",
           "test_jacobian.jl",
           "test_jacobian_extended.jl",
-          "test_adjoint.jl",
           "test_gradient_fwi.jl",
-          "test_all_options.jl",
           "test_gradient_lsrtm.jl"]
+        #   "test_gradient_twri.jl"]
```
Comment (alisiahkoohi) [inline test/runtests.jl:36]: should we uncomment this?

File: test/test_basics.jl (hunk 8 above)
Comment (mloubout) [reply test/test_basics.jl:58]: Looks like some leftover duplicate indeed, removed from `test_composite.jl`

File: test/runtests.jl (hunk 9 above)
Comment (mloubout) [reply test/runtests.jl:36]: No this fails most of the time. And talking with @grizzuti he's ok commenting it out since the gradient seems fine by itself.

File: test/runtests.jl [hunk 10]
```diff
@@ -69,16 +78,17 @@ if GROUP == "TTI_OP" || GROUP == "All"
...
-        include(t)
+        @time include(t)
     end
 end
 
 # Anisotropic Acoustic tests with free surface
 if GROUP == "TTI_OP_FS" || GROUP == "All"
-    println("JUDI TTI operators with free surfacetests")
+    println("JUDI TTI operators with free surface tests")
+    deleteat!(devito, devito .== "test_gradient_twri.jl")
```
Comment (alisiahkoohi) [inline test/runtests.jl:88]: Why is this necessary?

Comment (mloubout) [reply test/runtests.jl:88]: Leftover from when `test_gradient_twri` was still run, removed.

**Review by alisiahkoohi — APPROVED** [review]
Looks good, apart from a few small points described in the comments. Once those are addressed it is ready for merge in my opinion.

**Comment by mloubout** [comment]
>Once those are addressed it is ready for merge in my opinion.

All fixed, thanks for going through it
